### PR TITLE
Kyuubi Spark External Shuffle Service

### PR DESCRIPTION
**Pull request by yaooqinn**
#### Kyuubi Spark External Shuffle Service

This enables the missing feature of Dynamic Resource Allocation(a.k.a, DRA, DA) for running Spark on Kubernetes, which provides the auto-scaling capability for Spark application, especially downscaling.

Although there is a feature called shuffle tracking to track the shuffle data and tell Spark to remove executor while the data exceeds the time to live, these removals are coarse-grained.

##### Prerequisites

- Everything you need to know about Spark on Kubernetes
- Running Spark on Kubernetes apps with hostNetwork enabled
- Running this module as a DaemonSet with hostNetwork enabled
- Self-managed shuffle data cleaner as the state of apps will not be posted to this service
  - See: tools/spark-block-cleaner/kubernetes/spark-block-cleaner.yml

##### Usage

Check `tools/kyuubi-shuffle-service/kubernetes/example.yaml` for more information.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

#### cmd
```
bin/spark-sql 
```

#### config
```
spark.kubernetes.file.upload.path=./
spark.master=k8s://https://kubernetes.docker.internal:6443
spark.kubernetes.container.image=yaooqinn/spark:v20210619
spark.kubernetes.context=docker-for-desktop_1
spark.dynamicAllocation.enabled=true
spark.dynamicAllocation.minExecutors=2
spark.dynamicAllocation.maxExecutors=10
spark.driver.memory 5g
spark.kubernetes.executor.podNamePrefix=sparksql
spark.shuffle.service.enabled=true

spark.kubernetes.executor.podTemplateFile=/Users/kentyao/Downloads/executor-template.yml
spark.kubernetes.executor.volumes.hostPath.spark-local-dir-1.mount.path=/opt/spark/data1

spark.kubernetes.executor.volumes.hostPath.spark-local-dir-1.options.path=/Users/kentyao/Downloads/data/1

spark.kubernetes.executor.volumes.hostPath.spark-local-dir-2.mount.path=/opt/spark/data2

spark.kubernetes.executor.volumes.hostPath.spark-local-dir-2.options.path=/Users/kentyao/Downloads/data/2
spark.blockManager.port=0
```

#### query
```
SELECT /*+ REPARTITION */ 1 as pid, id from range(1, 10000, 1, 500);
```

### CODE REVIEW

File: tools/kyuubi-shuffle-service/kubernetes/example.yaml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+              mountPath: /opt/spark/data2
+          env:
+            # Change this in case of OOM based on you spec.containers.resources.* for
+            # the heap memory of shuffle service proc
+            - name: SPARK_DAEMON_MEMORY
+              value: 512M
+            # SPARK_SHUFFLE_OPTS allows us to pass spark configurations to the shuffle service
+            # spark.local.dir points to spec.containers.volumeMounts.*, shall be same as Spark
+            # app side.
+            # spark.shuffle.cleaner.interval=30s, this currently not work because only can a
```
Comment (yaooqinn) [inline tools/kyuubi-shuffle-service/kubernetes/example.yaml:57]: If an `ExternalBlockStoreClient` can also post ShuffleServiceHeartbea, we can reach much more fine grained disk/shuffle data management

**Comment by yaooqinn** [comment]
cc @holdenk @dongjoon-hyun to see if we have an opportunity to make the current external shuffle service work for all cluster managers, including k8s, standalone, etc.

**Comment by holdenk** [comment]
So I don't know anything about this project but certainly does seem like it could be interesting. This provides an alternative shuffle service for Spark to use on Kube? Is there a design doc? I notice hostNetwork has to be set to true which is something I'm generally speaking not comfortable with for security reasons, is that a temporary design decision or something more permanent?

**Comment by yaooqinn** [comment]
Thanks @holdenk 

> So I don't know anything about this project but certainly does seem like it could be interesting. 

Do worry about the project itself. The project relies on Spark who can achieve better scalability to use computing resources more effectively on Kubernetes. And this is the motivation of this PR.

> This provides an alternative shuffle service for Spark to use on Kube? Is there a design doc?

Yes, this provides an external shuffle service for Spark to use on Kube. The current implementation is codeless as you can see in this PR. We only need to write a Dockerfile(3 LOC only) for the shuffle service(`org.apache.spark.deploy.ExternalShuffleService`) endpoint based on the official Spark image and then deploy it on k8s as a DaemonSet. This can work with official releases(verified 3.1.2) without any modification. The only thing a bit tricky here is that users need to enable the `hostNetwork` to let the client successfully establish the connection `LOCAL` shuffle server during an executor initialization.

> I notice hostNetwork has to be set to true which is something I'm generally speaking not comfortable with for security reasons, is that a temporary design decision or something more permanent?

I agree that using hostNetwork is not secure and incomprehensive for this feature. I will try to figure out a solution with the Kubernetes pod network only, maybe in the next few weeks with a SPIP or a POC PR in the Spark community.

**Comment by holdenk** [comment]
I like the simplicity of this approach, but I think there are some pretty serious limitations we should think through. It would let us scale up and down executor pods but not the shuffle services and we'd probably end up doing a lot of remote fetches. Can you give me some more context around your plans here? Have you looked at the current work on introducing the new shuffle APIs for Kube in Spark?

Is the goal just to support dynamic scaling? If so have you tried enabling migrations?

**Comment by jerqi** [comment]
@yaooqinn Is there any progress or further discuss about K8S ESS?

**Pull request closed without merging**